nucypher/blockchain/eth/interfaces.py

```diff
@@ -267,18 +269,28 @@ def attach_middleware(self):
         if self.poa is None:  # If POA is not set explicitly, try to autodetect from chain id
             chain_id = int(self.client.chain_id)
             self.poa = chain_id in POA_CHAINS
-            self.log.debug(f'Autodetecting POA chain ({self.client.chain_name})')
+
+        self.log.debug(f'Ethereum chain: {self.client.chain_name} (chain_id={chain_id}, poa={self.poa})')
 
         # For use with Proof-Of-Authority test-blockchains
         if self.poa is True:
             self.log.debug('Injecting POA middleware at layer 0')
             self.client.inject_middleware(geth_poa_middleware, layer=0)
 
-        # Gas Price Strategy
-        self.client.w3.eth.setGasPriceStrategy(self.gas_strategy)
-        self.client.w3.middleware_onion.add(middleware.time_based_cache_middleware)
-        self.client.w3.middleware_onion.add(middleware.latest_block_based_cache_middleware)
-        self.client.w3.middleware_onion.add(middleware.simple_cache_middleware)
+        # Gas Price Strategy:
+        # Bundled web3 strategies are too expensive for Infura (it takes ~1 minute to get a price),
+        # so we use external gas price oracles, instead (see #2139)
+        if isinstance(self.client, InfuraClient):
+            gas_strategy = oracle_fallback_gas_price_strategy
+        else:
+            gas_strategy = self.gas_strategy
+        self.client.set_gas_strategy(gas_strategy=gas_strategy)
+        gwei_gas_price = Web3.fromWei(self.client.generate_gas_price(), 'gwei')
+        self.log.debug(f"Currently, our gas strategy returns a gas price of {gwei_gas_price} gwei")
```

Comment:
Is it possible to add gas strategy name here? Because overwriting for Infura clients could confuse for the first time

Reply:
The problem is that gas strategies themselves don't have names, as they're plain functions.

Reply:
I guess you could use `.__name__`. Or functors instead of functions if names are really important.